jupyterhub_carina/CarinaSpawner.py

```diff
@@ -0,0 +1,218 @@
+import docker
+from dockerspawner import DockerSpawner
+import os.path
+import re
+import requests
+import shutil
+from tornado import gen
+from traitlets import Dict, Integer, Unicode
+from .CarinaOAuthClient import CarinaOAuthClient
+
+
+class CarinaSpawner(DockerSpawner):
+    """
+    Spawn the user's Jupyter server on their Carina cluster.
+    """
+
+    cluster_name = Unicode(
+        'jupyterhub',
+        help="The name of the user's Carina cluster.",
+        config=True)
+
+    # Override the default container name. Since we are running on the user's cluster, we don't need the username suffix
+    container_name = Unicode(
+        'jupyter',
+        help="The name of the Jupyter server container running on the user's cluster.",
+        config=True)
+
+    cluster_polling_interval = Integer(
+        30,
+        help="The number of seconds between polling for a user's cluster to become active.",
+        config=True
+    )
+
+    # Override the default timeout to allow extra time for creating the cluster and pulling the server image
+    start_timeout = Integer(
+        300,
+        help=DockerSpawner.start_timeout.help,
+        config=True)
+
+    # Override the docker run parameters
+    extra_host_config = Dict(
+        {
+            'volumes_from': ['swarm-data'],  # --volumes-from swarm-data
+            'port_bindings': {8888: None},   # -p 8888:8888
+            'restart_policy': {              # --restart always
+                'MaximumRetryCount': 0,
+                'Name': 'always'
+            },
+        },
+        help=DockerSpawner.extra_host_config.help,
+        config=True)
+
+    def __init__(self, **kwargs):
+        # Use a different docker client for each server
+        self._client = None
+        self._carina_client = None
+        
+        super().__init__(**kwargs)
+
+    @property
+    def client(self):
+        """
+        The Docker client used to connect to the user's Carina cluster
+        """
+
+        # TODO: Figure out how to configure this without overriding, or tweak a bit and call super
```

Comment:
Also I'm not sure if you noticed, but because each spawner's Docker client uses unique credentials, I am using an instance variable for the client, instead of the class variable like DockerSpawner. So it's not just the client creation logic that is different.